plugins/database/src/db_plugin.cpp

```diff
@@ -15599,6 +15599,60 @@ auto db_data_object_finalize_op(irods::plugin_context& _ctx, const char* _json_i
     return SUCCESS();
 } // db_data_object_finalize_op
 
+auto db_check_auth_credentials_op(irods::plugin_context& _ctx,
+                                  const char* _username,
+                                  const char* _zone,
+                                  const char* _password,
+                                  int* _would_succeed) -> irods::error
+{
+    if (const auto ret = _ctx.valid(); !ret.ok()) {
+        return PASS(ret);
+    }
+
+    if (!_username || !_zone || !_password || !_would_succeed) {
+        log_db::error("{}: Received one or more null pointers.", __func__);
+        return ERROR(SYS_INVALID_INPUT_PARAM, "Received one or more null pointers.");
+    }
+
+    *_would_succeed = -1; // Indicates the correctness of the creds is unknown.
+
+    std::array<char, MAX_PASSWORD_LEN + 20> decoded_password{};
+
+    if (const auto ec = decodePw(_ctx.comm(), _password, decoded_password.data()); ec < 0) {
+        log_db::error("{}: Failed to decode password with error code [{}].", __func__, ec);
+        return ERROR(ec, "Password decode error.");
+    }
+
+    icatScramble(decoded_password.data());
+
+    try {
+        auto [db_instance, db_conn] = irods::experimental::catalog::new_database_connection();
+
+        nanodbc::statement stmt{db_conn};
+        nanodbc::prepare(stmt,
+                         "select u.user_id from r_user_main u "
+                         "inner join r_user_password p on u.user_id = p.user_id "
+                         "where u.user_name = ? and u.zone_name = ? and p.rcat_password = ?");
+
+        stmt.bind(0, _username);
+        stmt.bind(1, _zone);
+        stmt.bind(2, decoded_password.data());
+
+        if (auto row = nanodbc::execute(stmt); !row.next()) {
+            log_db::warn("{}: Incorrect credentials for user [{}#{}].", __func__, _username, _zone);
+            *_would_succeed = 0;
+        }
+        else {
+            *_would_succeed = 1;
+        }
+
+        return SUCCESS();
+    }
+    catch (const std::exception& e) {
+        return ERROR(SYS_LIBRARY_ERROR, e.what());
+    }
```

Comment:
Not that this would happen, but should we cover the `catch (...)` case as well?

Reply:
I always flip flop on that. I don't think it is necessary, but you never know.

Will add.

Reply:
Done